File: Sequencial.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;

public class Sequencial {

    public static String lerArquivo(String caminho) throws IOException {
        return new String(Files.readAllBytes(Paths.get(caminho)));
    }

    public static void salvarArquivo(String caminho, String conteudo) throws IOException {
        Files.write(Paths.get(caminho), conteudo.getBytes(), StandardOpenOption.CREATE);
    }

    public static void main(String[] args) {
        String nomeDoArquivo = "arquivos/DNA.txt";
        String nomeDoArquivoSaida = "arquivos/DNA_transcrito.txt";

        String conteudo;
        try {
            conteudo = lerArquivo(nomeDoArquivo);
        } catch (IOException e) {
            System.out.println("Erro ao abrir o arquivo!");
            return;
        }

        char[] fita = conteudo.toCharArray();

        long tempoInicial = System.currentTimeMillis();

        for (int i = 0; i < fita.length; i++) {
            switch (Character.toUpperCase(fita[i])) {
                case 'A':
                    fita[i] = 'T';
                    break;
                case 'C':
                    fita[i] = 'G';
                    break;
                case 'T':
                    fita[i] = 'A';
                    break;
                case 'G':
                    fita[i] = 'C';
                    break;
            }
        }

        long tempoFinal = System.currentTimeMillis();
        long tempoDecorrido = tempoFinal - tempoInicial;

        System.out.printf("Tempo decorrido: %d ms%n", tempoDecorrido);

        try {
            salvarArquivo(nomeDoArquivoSaida, new String(fita));
            System.out.println("Arquivo transcrito salvo em: " + nomeDoArquivoSaida);
        } catch (IOException e) {
            System.out.println("Erro ao salvar o arquivo transcrito!");
        }
    }
}

File: Sequencial.cpp
#include <iostream>
#include <unistd.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <omp.h>
#include <time.h>
#include <string.h>
#include <fstream>

clock_t tInicio, tFim, tDecorrido;
using std::cout; using std::string; using std::endl;
using std::ofstream;

string lerArquivo(const string& caminho) {
    struct stat sb{};
    string resultado;

    FILE* arquivo = fopen(caminho.c_str(), "r");
    if (!arquivo) {
        cout << "Erro ao abrir o arquivo!" << endl;
        return "";
    }

    stat(caminho.c_str(), &sb);
    resultado.resize(sb.st_size);
    fread(const_cast<char*>(resultado.data()), sb.st_size, 1, arquivo);
    fclose(arquivo);

    return resultado;
}

void salvarArquivo(const string& caminho, const char* conteudo) {
    ofstream arquivoSaida(caminho);
    if (!arquivoSaida) {
        cout << "Erro ao criar o arquivo!" << endl;
        return;
    }
    arquivoSaida << conteudo;
    arquivoSaida.close();
}

int main() {
    string nomeDoArquivo("arquivos/DNA.txt");
    string conteudo = lerArquivo(nomeDoArquivo);
    
    if (conteudo.empty()) {
        return -1; 
    }
    
    char fita[conteudo.length() + 1];
    strcpy(fita, conteudo.c_str());

    tInicio = clock();
    for (int i = 0; i < conteudo.length(); i++) {		
        switch (toupper(fita[i])) {
            case 'A':
                fita[i] = 'T';
                break;
            case 'C':
                fita[i] = 'G';
                break;
            case 'T':
                fita[i] = 'A';
                break;
            case 'G':
                fita[i] = 'C';
                break;
        }
    }
    tFim = clock();

    tDecorrido = ((tFim - tInicio) * 1000 / CLOCKS_PER_SEC); 
    printf("Tempo decorrido: %ld ms\n", tDecorrido);

    string nomeDoArquivoSaida("arquivos/DNA_transcrito.txt");
    salvarArquivo(nomeDoArquivoSaida, fita);

    return 0;
}
